### PR TITLE
Fixed blurring empty image base caption

### LINKED ISSUE #1776: Empty placeholder not hidden when blurred
## Are you reporting a feature request or a bug?

Bug

## Provide detailed reproduction steps (if any)

1. Open http://tests.ckeditor.test:1030/tests/plugins/imagebase/features/manual/caption
2. Focus widget with empty caption.
3. Put some content into the empty caption.
4. Focus widget with non-empty caption (the last one in the editor).
5. Select all content inside the caption and delete it.
6. Blur the widget.

### Expected result

Caption for the last widget is hidden.

### Actual result

Caption is still visible, despite the fact it's empty.

## Other details

* CKEditor version: 4.9.0
* Installed CKEditor plugins: `easyimage`

### PR DESCRIPTION
**Pull request by jacekbogdanski**
## What is the purpose of this pull request?

Bug fix.

## Does your PR contain necessary tests?

All patches which change the editor code must include tests. You can always read more
on [PR testing](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_contributing_code-section-tests),
[how to set the testing environment](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_tests) and
[how to create tests](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_tests-section-creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [x] Unit tests
- [x] Manual tests

## What changes did you make?

Changed `imagebase` plugin to refresh every caption featured widget on `selectionChange`.

Closes #1776

### CODE REVIEW

File: CHANGES.md [hunk 1]
```diff
@@ -1,6 +1,12 @@
 ﻿CKEditor 4 Changelog
 ====================
 
+## CKEditor 4.9.1
```
Comment (Comandeer) [inline CHANGES.md]: Please fix version – as 4.9.1 is dropped, it should be 4.10.0

File: plugins/imagebase/plugin.js [hunk 2]
```diff
@@ -614,6 +608,16 @@
 		link: getLinkFeature()
 	};
 
+	function getWidgetsWithFeature( widgets, feature ) {
```
Comment (Comandeer) [inline plugins/imagebase/plugin.js]: Functions should be declared before they are used in the code.

File: plugins/imagebase/plugin.js [hunk 3]
```diff
@@ -614,6 +608,16 @@
...
+		return CKEDITOR.tools.array.reduce( CKEDITOR.tools.objectKeys( widgets ), function( featuredWidgets, id ) {
```
Comment (Comandeer) [inline plugins/imagebase/plugin.js]: Shouldn't it be `filter`, not `reduce`? Then the whole function could be simplified to:

```javascript
return hasWidgetFeature( widgets[ id ], feature );
```

File: tests/plugins/imagebase/features/manual/captionemptyplaceholder.md [hunk 4]
```diff
@@ -0,0 +1,15 @@
+@bender-tags: 4.9.1, bug, 1776
```
Comment (Comandeer) [inline tests/plugins/imagebase/features/manual/captionemptyplaceholder.md]: Please update version tag.

File: tests/plugins/imagebase/features/manual/captionemptyplaceholder.md [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+@bender-ui: collapsed
+@bender-ckeditor-plugins: wysiwygarea, toolbar, basicstyles, imagebase
+@bender-include: %BASE_PATH%/plugins/easyimage/_helpers/tools.js
+
+1. Focus first image.
```
Comment (Comandeer) [inline tests/plugins/imagebase/features/manual/captionemptyplaceholder.md]: I'm not sure if this procedure is enough to reproduce the bug. It seems that typing and deleting – as stated in #1776 - is necessary.

File: plugins/imagebase/plugin.js (hunk 3 above)
Comment (jacekbogdanski) [reply plugins/imagebase/plugin.js]: I'm not sure how to write it as a filter function? `editor.widgets.instances` is an object so we will have to convert it to an array to use filter function which requires additional code. If I simply replace `reduce` with `filter` here I will get widget ids, not instances. WDYT @Comandeer ?

File: tests/plugins/imagebase/features/manual/captionemptyplaceholder.md (hunk 5 above)
Comment (jacekbogdanski) [reply tests/plugins/imagebase/features/manual/captionemptyplaceholder.md]: The HTML of this manual test already have empty `figcaption` element to reproduce removing caption content. I reproduced manual test steps without my changes and the test was broken so I think it works fine. I created it to shorten test steps but if you think it's too terse I will refactor the test.

File: CHANGES.md (hunk 1 above)
Comment (jacekbogdanski) [reply CHANGES.md]: I updated it to `4.10` to keep consistent with the rest of the file.

File: plugins/imagebase/plugin.js (hunk 3 above)
Comment (Comandeer) [reply plugins/imagebase/plugin.js]: True, I didn't think about it. So let's stay with `reduce` as it's the best option here.

File: tests/plugins/imagebase/features/manual/captionemptyplaceholder.md (hunk 5 above)
Comment (Comandeer) [reply tests/plugins/imagebase/features/manual/captionemptyplaceholder.md]: Strangely, I could reproduce it in your test, but no success in `/tests/plugins/imagebase/features/manual/caption` 🤔

File: plugins/imagebase/plugin.js (hunk 2 above)
Comment (jacekbogdanski) [reply plugins/imagebase/plugin.js]: I see now that functions in this file are declared at the top and I moved my function also. But are we going with this code style in whole CKEditor codebase? Linter didn't complain about it and I think it's cleaner when your private functions are below the main login. Also I didn't catch this code style in different files. Should I follow this rule in other files?

**Pull request merged**